azurerm/internal/services/storage/storage_management_policy_resource.go

```diff
@@ -243,172 +311,334 @@ func resourceStorageManagementPolicyDelete(d *schema.ResourceData, meta interfac
+			},
 		}
+		if (rule.Definition.Actions.Version != nil || rule.Definition.Actions.Snapshot != nil) && rule.Definition.Filters.BlobIndexMatch != nil {
+			return nil, fmt.Errorf("`blob_index_match_tag` is not supported as a filter for versions and snapshots")
+		}
+		result = append(result, rule)
 	}
 	return &result, nil
 }
 
-func expandStorageManagementPolicyRule(d *schema.ResourceData, ruleIndex int) storage.ManagementPolicyRule {
-	name := d.Get(fmt.Sprintf("rule.%d.name", ruleIndex)).(string)
-	enabled := d.Get(fmt.Sprintf("rule.%d.enabled", ruleIndex)).(bool)
-	typeVal := "Lifecycle"
-
-	definition := storage.ManagementPolicyDefinition{
-		Filters: &storage.ManagementPolicyFilter{},
-		Actions: &storage.ManagementPolicyAction{},
-	}
-	filtersRef := d.Get(fmt.Sprintf("rule.%d.filters", ruleIndex)).([]interface{})
-	if len(filtersRef) == 1 {
-		if filtersRef[0] != nil {
-			filterRef := filtersRef[0].(map[string]interface{})
-
-			prefixMatches := []string{}
-			prefixMatchesRef := filterRef["prefix_match"].(*schema.Set)
-			if prefixMatchesRef != nil {
-				for _, prefixMatchRef := range prefixMatchesRef.List() {
-					prefixMatches = append(prefixMatches, prefixMatchRef.(string))
-				}
-			}
-			definition.Filters.PrefixMatch = &prefixMatches
-
-			blobTypes := []string{}
-			blobTypesRef := filterRef["blob_types"].(*schema.Set)
-			if blobTypesRef != nil {
-				for _, blobTypeRef := range blobTypesRef.List() {
-					blobTypes = append(blobTypes, blobTypeRef.(string))
-				}
-			}
-			definition.Filters.BlobTypes = &blobTypes
+func expandStorageManagementPolicyFilters(inputs []interface{}) *storage.ManagementPolicyFilter {
+	if len(inputs) == 0 {
+		return nil
+	}
+	input := inputs[0].(map[string]interface{})
+
+	return &storage.ManagementPolicyFilter{
+		PrefixMatch:    utils.ExpandStringSlice(input["prefix_match"].(*schema.Set).List()),
+		BlobTypes:      utils.ExpandStringSlice(input["blob_types"].(*schema.Set).List()),
+		BlobIndexMatch: expandAzureRmStorageBlobIndexMatch(input["blob_index_match_tag"].(*schema.Set).List()),
+	}
+}
+
+func expandStorageManagementPolicyActions(inputs []interface{}) *storage.ManagementPolicyAction {
+	if len(inputs) == 0 {
+		return nil
+	}
+	input := inputs[0].(map[string]interface{})
+
+	return &storage.ManagementPolicyAction{
+		BaseBlob: expandStorageManagementPolicyActionsBaseBlob(input["base_blob"].([]interface{})),
+		Snapshot: expandStorageManagementPolicyActionsSnapshot(input["snapshot"].([]interface{})),
+		Version:  expandStorageManagementPolicyActionsVersion(input["version"].([]interface{})),
+	}
+}
+
+func expandStorageManagementPolicyActionsBaseBlob(inputs []interface{}) *storage.ManagementPolicyBaseBlob {
+	if len(inputs) == 0 {
+		return nil
+	}
+	input := inputs[0].(map[string]interface{})
+	result := storage.ManagementPolicyBaseBlob{}
+
+	if v, ok := input["delete_after_days_since_modification_greater_than"].(float64); ok && v != -1 {
```

Comment:
this means this value can't be removed, is that intentional?

Reply:
same above?